### PR TITLE
Support function calls for IF statements

### PR DESCRIPTION
**Pull request by guipublic**
This PR adds support for IF statements and function calls:
- If statement can be added inside a function body: the function CFG is reduced to take into account IF statements
- a function/intrinsic can be called inside a IF branch: we add a predicate to the function call which is used during inlining to 'conditionalise' instructions that have potential side-effects.
For intrinsic we simply 'conditionalise' their returned arrays and we assume intrinsics can never fail, which must still be verified.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/inline.rs [hunk 1]
```diff
@@ -227,56 +265,100 @@ pub fn inline_in_block(
...
-                        ctx.get_or_create_const(FieldElement::from(offset as i128), index_type);
-
-                    let add =
-                        node::Binary { operator: node::BinaryOp::Add, lhs: offset_id, rhs: *index };
-                    let adr_id = ctx.new_instruction(Operation::Binary(add), index_type)?;
+                    let adr_id = if *index == NodeId::dummy() {
+                        NodeId::dummy()
+                    } else {
+                        let index_type = ctx[*index].get_type();
+                        let offset_op = if offset < 0 {
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/inline.rs:293]: Why is this additional negative offset branch needed? New functionality or bugfix?

File: crates/noirc_evaluator/src/ssa/node.rs [hunk 2]
```diff
@@ -521,8 +531,12 @@ impl Instruction {
...
 pub enum Operation {
     Binary(Binary),
-    Cast(NodeId),                                                 //convert type
-    Truncate { value: NodeId, bit_size: u32, max_bit_size: u32 }, //truncate
+    Cast(NodeId), //convert type
+    Truncate {
+        value: NodeId,
+        bit_size: u32,
+        max_bit_size: u32,
+    }, //truncate
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/node.rs:539]: Looks like `use_small_heuristics = "Max"` in our .rustfmt.toml broke.
Looking at https://rust-lang.github.io/rustfmt I no longer see it as an option so I wondered if it was removed but it doesn't seem to be removed in the rustfmt repo.

File: crates/nargo/tests/test_data/9_conditional/src/main.nr [hunk 3]
```diff
@@ -1,8 +1,67 @@
...
+                let c = a[j];
+                a[j] = a[j+1];
+                a[j+1]= c;
+            }
+        };
+    };
+    a
+}
+
+fn call_intrinsic(x : [5]u8, result : pub [32]u8) {
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/nargo/tests/test_data/9_conditional/src/main.nr:17]: Yes, it came from a copy/paste, so disallowing it is a good idea.

File: crates/noirc_evaluator/src/ssa/inline.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/inline.rs:293]: It's a bugfix, basically the safeguard I added in the constant folding (in the `wrapping()` function) are not compatible with unsigned numbers. This should be fixed as well.
In any cases, this offset stuff is becoming annoying. I will remove it and use an absolute addressing once I am done with the IFs.

File: crates/noirc_evaluator/src/ssa/function.rs [hunk 4]
```diff
@@ -231,8 +251,8 @@ pub fn create_function(
     }
     igen.context
         .new_instruction(node::Operation::Return(returned_values), node::ObjectType::NotAnObject)?;
-    func.compile(igen)?; //unroll the function
-
+    let decision = func.compile(igen).unwrap(); //unroll the function
```
[the remark this answers is not in the document]
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/function.rs:254]: No clue, I added it back.

**Review by jfecher — APPROVED** [review]

**Pull request merged**